adds agrocd integration dockerfile, scripts, doc  and examples (#732)
accurics/terrascan: docs/integrations/argocd-integration.md

``` 

After making changes to the webhook deployment file, apply this yaml in your cluster.

You can also run terrascan admission controller server outside cluster, for more information on configuring terrascan as an admission controller webhook, follow https://docs.accurics.com/projects/accurics-terrascan/en/latest/integrations/admission-controller-webhooks-usage 

#### Step 2: Create a Dockerfile for the container which has the terrascan script to run the remote scan against the terrascan’s admission controller webhook.

```DockerFile
# Dockerfile with a script to use terrascan's validating webhook
# configured in the kubernetes cluster, to scan a repo for violations
FROM alpine:3.12.0

#curl to send request to terrascan validating webhook
RUN apk add --no-cache curl 

WORKDIR /home/terrascan

RUN mkdir bin

COPY scripts/argocd-terrascan-remote-scan.sh  bin/terrascan-remote-scan.sh

# create non root user
RUN addgroup --gid 101 terrascan && \
  adduser -S --uid 101 --ingroup terrascan terrascan

RUN chown -R terrascan:terrascan bin && \
  chmod u+x bin/terrascan-remote-scan.sh

USER terrascan

CMD ["sh"]
```

terrascan-remote-scan script

```sh
#!/bin/sh

> TERRASCAN_SERVER="https://${SERVICE_NAME}"

Reviewer: -we should make the script more robust.
-we can handle cases where some required env variables are not provided. (we should print a message and exit non zero)
-also, we should make the vars configurable, so that (if required) we would be able to extend the scan to other iac's and types as well

Author: added check for required env variables.

IAC_VERSION="v1"
CLOUD_PROVIDER="all"
REMOTE_TYPE="git"

SCAN_URL="${TERRASCAN_SERVER}/v1/${IAC}/${IAC_VERSION}/${CLOUD_PROVIDER}/remote/dir/scan"

RESPONSE=$(curl -s -w \\n%{http_code} --location -k  --request POST "$SCAN_URL" \
--header 'Content-Type: application/json' \
--data-raw '{
"remote_type":"'${REMOTE_TYPE}'",
"remote_url":"'${REMOTE_URL}'"
}')

echo "$RESPONSE"

HTTP_STATUS=$(printf '%s\n' "$RESPONSE" | tail -n1)

if [ "$HTTP_STATUS" -ne 200 ]; then
  exit 3
fi
```

#### Step 3: Configure PreSync hook to use container created in step 2

The following example hook yaml is mostly ready to be added to an existing kubernetes configuration.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
generateName: terrascan-hook-
namespace: test
annotations:
  argocd.argoproj.io/hook: PreSync            
spec:
ttlSecondsAfterFinished: 3600
template:
  spec:
    securityContext:
      seccompProfile:
        type: RuntimeDefault
    containers:
    - name: terrascan-argocd
      image: <IMAGE FROM STEP 2>
      resources:
        requests:
          cpu: "1"
          memory: "256Mi"
        limits:
          cpu: "1"
          memory: "256Mi"
      env:
        - name: SERVICE_NAME

>           value: terra-controller-service.default.svc

Reviewer: this value would be name of the service used to expose admission controller webhook.

## Terrascan can be integrated with Argo CD in two ways
___
1. Use terrascan as a pre-sync hook to scan remote repositories
2. Use terrascan’s k8s admission controller along with a pre-sync that scans a configured repository with the admission controller webhook

### 1. Configure terrascan as a PreSync hook and scan the remote repository. 
___

#### Configure a PreSync hook

The following example hook yaml is mostly ready to be added to an existing kubernetes configuration. Just make sure that the secrets,  known_hosts  and ssh_config volume are relevant and specify a terrascan image. You can also map a slack notification script to the container which will send notifications to your Slack webhook endpoint after the embedded script scans the repo.
 
```yaml 
apiVersion: batch/v1
kind: Job
metadata:
 generateName: terrascan-hook-
 annotations:
   argocd.argoproj.io/hook: PreSync
spec:
 ttlSecondsAfterFinished: 3600
 template:
   spec:
     securityContext:
       seccompProfile:
         type: RuntimeDefault
     # if you want to add slack notification script add one more volume here
     Volumes:
       - name: notification-scripts

>          hostPath:

Reviewer 2: Can you change this to use configMap ?

Author: changed to configmap

Config-map: 

``` 
  kubectl  create configmap ssh-known-hosts --from-file=
   < path to you known hosts file >
```   
 
ssh config secret

```
 kubectl create secret generic ssh-config-secret \
   --from-file=< path to your ssh config file >
```   
 
Example ssh config file

``` 
 Host github.com
  HostName github.com
  IdentityFile ~/.ssh/id_ed25519_github
```

Once the presynchook yaml file is completely configured, add this file to your repository folder for which Argo CD pipeline is configured.

### 2. Use Terrascan Admission Webhook from PreSyncHook
___
You can use the already deployed terrascan’s k8s admission controller webhook to scan the remote repository from Argo CD PreSync hook.
To configure, follow below steps

#### Step 1: Configure terrascan admission controller webhook deployment yaml file with required keys and volumes and service to expose the controller pod.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
name: terra-controller-webhook
labels:
  app: terrascan-admission-webhook
spec:
replicas: 1
selector:
  matchLabels:
    app: terrascan-admission-webhook
template:
  metadata:
    labels:
      app: terrascan-admission-webhook
  spec:
    containers:

>     - name: terrascan-admission-webhook

Reviewer 2: ```suggestion
    - name: terrascan
```

Author: This is just name for the pod it doesnot have any impact user can configure it as required

Reviewer 2: yes but it can confuse the reader/user.

>       image: <TERRASCAN IMAGE>

Reviewer 2: accurics/terrascan:1.5.1

Reviewer 2: We'll need to use a new tag from a new release that has the changes from the recents PRs including #724

Author: Yes it user can use any latest image the required changes are already done for integration in main terrascan image

      livenessProbe:
        initialDelaySeconds: 30
        periodSeconds: 10
        timeoutSeconds: 5
        httpGet:
          path: /health
          port: 443
          scheme: HTTPS
      env:
        - name: K8S_WEBHOOK_API_KEY
          value: yoursecretapikey
      volumeMounts:
        - mountPath: /data
          name: certspath
        - mountPath: /etc/github-secret
          name: ssh-keys-github
          readOnly: true
        - mountPath: /etc/ssh-config-secret
          name: ssh-config
          readOnly: true
        - mountPath: /etc/ssh-known-hosts-secret
          name: ssh-known-hosts
          readOnly: true
      command: ["/bin/sh", "-c"]
      args:
        - >
          cp /etc/github-secret/ssh-private-key /home/terrascan/.ssh/id_ed25519_github &&
          cp /etc/ssh-config-secret/ssh-config /home/terrascan/.ssh/config &&
          cp /etc/ssh-known-hosts-secret/ssh-known-hosts /home/terrascan/.ssh/known_hosts &&
          chmod -R 400 /home/terrascan/.ssh/* &&
          terrascan server --cert-path /data/server.crt --key-path /data/server.key -p 443 -l debug -c /data/config.toml
    volumes:
      - name: certspath
        hostPath:

>           path: <YOUR CERTIFICATE FOLDER PATH>

Reviewer 2: use a secret here instead?

Author: added changes to use secret instead of hostpath

WORKDIR /home/terrascan

RUN mkdir bin

COPY scripts/argocd-terrascan-remote-scan.sh  bin/terrascan-remote-scan.sh

# create non root user
RUN addgroup --gid 101 terrascan && \
  adduser -S --uid 101 --ingroup terrascan terrascan

RUN chown -R terrascan:terrascan bin && \
  chmod u+x bin/terrascan-remote-scan.sh

USER terrascan

CMD ["sh"]
```

terrascan-remote-scan script

```sh
#!/bin/sh
TERRASCAN_SERVER="https://${SERVICE_NAME}"
IAC="k8s"
IAC_VERSION="v1"
CLOUD_PROVIDER="all"
REMOTE_TYPE="git"

SCAN_URL="${TERRASCAN_SERVER}/v1/${IAC}/${IAC_VERSION}/${CLOUD_PROVIDER}/remote/dir/scan"

RESPONSE=$(curl -s -w \\n%{http_code} --location -k  --request POST "$SCAN_URL" \
--header 'Content-Type: application/json' \
--data-raw '{
"remote_type":"'${REMOTE_TYPE}'",
"remote_url":"'${REMOTE_URL}'"
}')

echo "$RESPONSE"

HTTP_STATUS=$(printf '%s\n' "$RESPONSE" | tail -n1)

if [ "$HTTP_STATUS" -ne 200 ]; then
  exit 3
fi
```

#### Step 3: Configure PreSync hook to use container created in step 2

The following example hook yaml is mostly ready to be added to an existing kubernetes configuration.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
generateName: terrascan-hook-

> namespace: test

Reviewer 2: don't hard-code the namespace

  while IFS= read line; do
          echo "${line}" | tr '\\"' ' ' >> results.out
  done
 
  cat results.out
 
  send_slack_notificaton $1 $2 $3
 
  echo "notification exit code: $?"
else
  echo "no response skipping"
fi
``` 
 
For non-public repositories, the private key, known_hosts and ssh config needs to be added as a kubernetes secret, configmap and secret respectively.
 
```
 kubectl create secret generic github-secret \
   --from-file=ssh-privatekey= < path to your private key > \
    --from-file=ssh-publickey=< path to your public key >
```

Config-map: 

``` 
  kubectl  create configmap ssh-known-hosts --from-file=< path to your known hosts file >
```   

``` 
  kubectl  create configmap slack-notifications --from-file=< path to your notification script >
```
 
ssh config secret

```
 kubectl create secret generic ssh-config-secret \
   --from-file=< path to your ssh config file >
```   
 
Example ssh config file

``` 
 Host github.com
  HostName github.com
  IdentityFile ~/.ssh/id_ed25519_github
```

Once the presynchook yaml file is completely configured, add this file to your repository folder for which Argo CD pipeline is configured.

> ### 2. Use Terrascan Admission Webhook from PreSyncHook

Reviewer 2: ```suggestion
### 2. Use PreSyncHook that sends scan requests to the existing Terrascan Server Service
```

___
You can use the already deployed terrascan’s k8s admission controller webhook to scan the remote repository from Argo CD PreSync hook.
To configure, follow below steps

> #### Step 1: Configure terrascan admission controller webhook deployment yaml file with required keys and volumes and service to expose the controller pod.

Reviewer 2: Rather than rendering these yamls , its better to point to a directory, guide or a blog ( Yusuf's blog for setting up the admission controller) . 

I have tried to create a crisp guide on setting up the terrascan server and webhook (if required) in my PR :
https://github.com/accurics/terrascan/pull/739/

With this:
1. We'll be able to follow DRY. 
2. We get a ready to use documentation with yaml files ready to be deployed to k8s.
3. Using kustomize helps in yaml manifests in a much neater way and doesn't really have a learning curve like helm.

set -o pipefail

TERRASCAN_SERVER="https://${SERVICE_NAME}"
IAC=${IAC_TYPE:-"k8s"}
IAC_VERSION=${IAC_VERSION:-"v1"}
CLOUD_PROVIDER=${CLOUD_PROVIDER:-"all"}
REMOTE_TYPE=${REMOTE_TYPE:-"git"}

SCAN_URL="${TERRASCAN_SERVER}/v1/${IAC}/${IAC_VERSION}/${CLOUD_PROVIDER}/remote/dir/scan"

RESPONSE=$(curl -s -w \\n%{http_code} --location -k  --request POST "$SCAN_URL" \
--header 'Content-Type: application/json' \
--data-raw '{
"remote_type":"'${REMOTE_TYPE}'",
"remote_url":"'${REMOTE_URL}'"
}')

echo "$RESPONSE"

HTTP_STATUS=$(printf '%s\n' "$RESPONSE" | tail -n1)

if [ "$HTTP_STATUS" -eq 403 ]; then
    exit 3
elif [ "$HTTP_STATUS" -eq 200 ]; then
    exit 0
else
    exit 1
fi
```

#### Step 3: Configure PreSync hook to use container created in step 2

The following example hook yaml is mostly ready to be added to an existing kubernetes configuration.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
generateName: terrascan-hook-
namespace: <YOUR APP NAMESPACE>
annotations:
  argocd.argoproj.io/hook: PreSync            
spec:
ttlSecondsAfterFinished: 3600
template:
  spec:
    securityContext:
      seccompProfile:
        type: RuntimeDefault
    containers:
    - name: terrascan-argocd

>       image: <IMAGE FROM STEP TWO>

Reviewer 2: we should provide a ready to use image here. a user shouldn't go through the hassle of build and push their own image to just checkout such features.

  done
 
  cat results.out
 
  send_slack_notificaton $1 $2 $3
 
  echo "notification exit code: $?"
else
  echo "no response skipping"
fi
``` 
 
For non-public repositories, the private key, known_hosts and ssh config needs to be added as a kubernetes secret, configmap and secret respectively.
 
```
 kubectl create secret generic ssh-key-secret \
   --from-file=ssh-privatekey= < path to your private key > \
    --from-file=ssh-publickey=< path to your public key >
```

Config-map: 

``` 
  kubectl  create configmap ssh-known-hosts --from-file=< path to your known hosts file >
```   

``` 
  kubectl  create configmap slack-notifications --from-file=< path to your notification script >
```
 
ssh config secret

```
 kubectl create secret generic ssh-config-secret \
   --from-file=< path to your ssh config file >
```   
 
Example ssh config file

``` 
 Host github.com
  HostName github.com
  IdentityFile ~/.ssh/id_ed25519_github
```

Once the presynchook yaml file is completely configured, add this file to your repository folder for which Argo CD pipeline is configured.

### 2. Use PreSyncHook to trigger the Terrascan Server Service
___

> You can use the already deployed terrascan’s k8s admission controller webhook to scan the remote repository from Argo CD PreSync hook.

Reviewer 2: ```suggestion
You can use the already deployed terrascan server service in k8s cluster to scan the remote repository from Argo CD PreSync hook.
```
